libraries/backend/azure_security_rules_helpers.rb

```diff
@@ -0,0 +1,300 @@
+      within_range = ip_range_check(destination_addresses_as_ip, criteria[:destination_ip_range])
+      return nil unless within_range
+      compliant = access == criteria[:access]
+      return compliant unless compliant
+    end
+
+    unless criteria[:source_service_tag].nil?
+      within_range = source_addresses_as_service_tag.include?(criteria[:source_service_tag])
+      return nil unless within_range
+      compliant = access == criteria[:access]
+      return compliant unless compliant
+    end
+
+    unless criteria[:destination_service_tag].nil?
+      within_range = destination_addresses_as_service_tag.include?(criteria[:destination_service_tag])
+      return nil unless within_range
+      compliant = access == criteria[:access]
+      return compliant unless compliant
+    end
+
+    # `Any`, `all` will fail.
+    # If this is the case, `protocol` parameter should not be provided at all.
+    unless criteria[:protocol].nil?
+      return nil unless criteria[:protocol].upcase == protocol
+      compliant = access == criteria[:access]
+      compliant = true if protocol == '*' && access == 'allow'
+      compliant = false if protocol == '*' && access == 'deny'
+      return compliant unless compliant
+    end
+
+    # Individual protocol number and/or a range can be provided.
+    #   ['8080', '50-60']
+    #   '22'
+    unless criteria[:source_port].nil?
+      ports = extract_ports(Array(criteria[:source_port]))
+      return nil unless ports.all? { |p| source_ports.include?(p) }
+      compliant = access == criteria[:access]
+      return compliant unless compliant
+    end
+
+    unless criteria[:destination_port].nil?
+      ports = extract_ports(Array(criteria[:destination_port]))
+      return nil unless ports.all? { |p| destination_ports.include?(p) }
+      compliant = access == criteria[:access]
+    end
+
+    compliant
+  end
+
+  private
+
+  # @return [Boolean]
+  # @param base_ip_ranges [Array] The list of IPAddr objects.
+  # @param criteria_ip_range [String] The IP range or address in CIDR format.
+  #
+  # criteria ip range will be checked whether it is within one of the base ip ranges.
+  #
+  def ip_range_check(base_ip_ranges, criteria_ip_range)
+    unless criteria_ip_range.match?(CIDR_IPV4_REG) || criteria_ip_range.match?(CIDR_IPV6_REG)
+      raise ArgumentError, 'IP range/address must be in CIDR format, e.g: `192.168.0.1/24, 2001:1234::/64`.'
+    end
+    criteria_ip_range_cidr = IPAddr.new(criteria_ip_range)
+    within_range = []
+    base_ip_ranges.each do |b|
+      within_range << b.include?(criteria_ip_range_cidr)
+    end
+    within_range.any? { |check| check == true }
+  end
+
+  # @return [Array] The list of ports as String type.
+  #   The range of valid port numbers if '*' provided.
+  # @param sources [Array, Class<Array>] The list of ports.
+  #
+  def extract_ports(sources)
+    return PORT_RANGE if sources.any? { |s| s == '*' }
+    sources.each_with_object([]) do |s, acc|
```

Comment:
similar to a previous comment. It would be good to have `acc` renamed to something more readable. There's a couple of instances of this in this file.